// fsm-engine/actions/actions.js
/**
 * Created by Ricardo Morais on 28/03/2017.
 */

let vm = require('vm');
let request = require('request');

/*
high lever actions arguments come in this format
    actionArguments: {
       argName1: value1,
       argName2: value2
    }
*/

let execute = function(action, sandbox, event, actionArguments) {

    try {

        let data = {
            action: action,
            arguments: {}
        };

        for (let key of Object.keys(actionArguments)) {
            if(key[0] != '$'){
                if (key.startsWith("expr")) {
                    let newKey = key.substring("expr".length);
                    newKey = newKey[0].toLowerCase() + newKey.substring(1);
                    data.arguments[newKey] = vm.runInContext(actionArguments[key], sandbox);
                } else {
                    data.arguments[key] = actionArguments[key];
                } 
            }
        }

        let host = actionArguments.exprHost ? vm.runInContext(actionArguments.exprHost, sandbox) : actionArguments.host;
        let route = "/execute";
        let errorEvent = actionArguments.errorEvent;
        let successEvent = actionArguments.successEvent;

        request({
            url: host + route,
            method: "POST",
            json: true,
            body: data,
        }, (error, response, body) => {
            if (error) {
                if (errorEvent) {
                    this.raise(errorEvent);
                }
                return;
            }
            if (successEvent) {
                this.raise(successEvent);
            }
        });

    } catch (err) {
        console.log(err);
    }
};

module.exports = execute;
